Update readme with consumers guide and features (#18)
jenkinsci/checks-api-plugin: README.md

[![GitHub Actions](https://github.com/jenkinsci/checks-api-plugin/workflows/CI/badge.svg?branch=master)](https://github.com/jenkinsci/checks-api-plugin/actions)
[![codecov](https://codecov.io/gh/jenkinsci/checks-api-plugin/branch/master/graph/badge.svg)](https://codecov.io/gh/jenkinsci/checks-api-plugin)

Inspired by the [GitHub Checks API](https://docs.github.com/en/rest/reference/checks#runs), this plugin aims to provide a general API to allow Jenkins plugins publishing checks (or reports) to remote source code management (SCM) platforms (e.g. GitHub, GitLab, BitBucket, etc.).

By consuming this API, other plugins can publish check with customized parameters for a Jenkins build, such as status, summary, warnings, code annotations, or even images.
Implementations of this API decide on how to make use of these parameters and where to publish the checks.

Known consumers:
* [Warnings Next Generation Plugin](https://plugins.jenkins.io/warnings-ng)
* [Code Coverage API Plugin](https://plugins.jenkins.io/code-coverage-api)

Implementations:
* [GitHub Checks Plugin](https://plugins.jenkins.io/github-checks)

> This plugin, alone with it's [GitHub implementation](https://plugins.jenkins.io/github-checks), has been installed on [ci.jenkins.io](https://ci.jenkins.io/Plugins) to help maintain hundreds of Jenkins plugins. You can take a look at the [action](https://github.com/jenkinsci/checks-api-plugin/runs/1025532156) for this repository or other plugin repositories under [Jenkins organization](https://github.com/jenkinsci) for the results.

Reviewer: ```suggestion
This plugin, along with it's [GitHub implementation](https://plugins.jenkins.io/github-checks), has been installed on [ci.jenkins.io](https://ci.jenkins.io/Plugins) to help maintain over 1500 Jenkins plugins. You can take a look at the [action](https://github.com/jenkinsci/checks-api-plugin/runs/1025532156) for this repository or other plugin repositories under [Jenkins organization](https://github.com/jenkinsci) for the results.
```

## Embedded Features

### Build Status Check

![GitHub Status](docs/images/github-status.png)

By listening to the Jenkins builds, this plugin will automatically publish statuses (pending, in progress, and completed) to different SCM platforms based on the remote repository the build is using.

### Pipeline Usage

Instead of depending on consumers plugins, the users can publish their checks directly in the pipeline script:

```
publishChecks name: 'example', title: 'Pipeline Check', summary: 'check through pipeline', text: 'you can publish checks in pipeline script', detailsURL: 'https://github.com/jenkinsci/checks-api-plugin#pipeline-usage'
```

## Guides

- [Consumers Guide](docs/consumers-guide.md)

> - [Implementation Guide](docs/implementation-guide.md)

Reviewer: doesn't appear to exist?

it should be added in the PR that adds it if I'm not misunderstanding

```suggestion
```

> ## Acknowledgements

Reviewer: probably worth adding the default sections from the archetype README,

Contributing and License?

e.g.
https://github.com/timja/jenkins-junit-postgresql-plugin#contributing

> This plugin is started as a [Google Summer of Code 2020 project](https://summerofcode.withgoogle.com/projects/#5139745388101632), you can find more about it on [Jenkins GSoC SIG](https://www.jenkins.io/sigs/gsoc/).

Reviewer: ```suggestion
This plugin was started as a [Google Summer of Code 2020 project](https://summerofcode.withgoogle.com/projects/#5139745388101632), you can find more about it on [Jenkins GSoC SIG](https://www.jenkins.io/sigs/gsoc/).
```

[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](https://issues.jenkins-ci.org/issues/?jql=component%20%3D%20checks-api-plugin)
[![Jenkins](https://ci.jenkins.io/job/Plugins/job/checks-api-plugin/job/master/badge/icon?subject=Jenkins%20CI)](https://ci.jenkins.io/job/Plugins/job/checks-api-plugin/job/master/)
[![GitHub Actions](https://github.com/jenkinsci/checks-api-plugin/workflows/CI/badge.svg?branch=master)](https://github.com/jenkinsci/checks-api-plugin/actions)
[![codecov](https://codecov.io/gh/jenkinsci/checks-api-plugin/branch/master/graph/badge.svg)](https://codecov.io/gh/jenkinsci/checks-api-plugin)

Inspired by the [GitHub Checks API](https://docs.github.com/en/rest/reference/checks#runs), this plugin aims to provide a general API to allow Jenkins plugins publishing checks (or reports) to remote source code management (SCM) platforms (e.g. GitHub, GitLab, BitBucket, etc.).

> By consuming this API, other plugins can publish check with customized parameters for a Jenkins build, such as status, summary, warnings, code annotations, or even images.

Reviewer: Let's add some line breaks to break up the paragraphs this is a new topic.

```suggestion

By consuming this API, other plugins can publish check with customized parameters for a Jenkins build, such as status, summary, warnings, code annotations, or even images.
```

> Then, the implementations of this API will decide on how to make use of these parameters and where to publish the checks.

Reviewer: ```suggestion
Implementations of this API decide on how to make use of these parameters and where to publish the checks.
```

> Current consumers of this plugin include [Warnings Next Generation Plugin](https://github.com/jenkinsci/warnings-ng-plugin) and [Code Coverage API Plugin](https://github.com/jenkinsci/code-coverage-api-plugin); current implementations include [GitHub Checks Plugin](https://github.com/jenkinsci/github-checks-plugin).

Reviewer: I think this looks better, might be nice to add integration pull requests so others can see how to add their own one

```suggestion
Known consumers:
* [Warnings Next Generation Plugin](https://plugins.jenkins.io/warnings-ng)
* [Code Coverage API Plugin](https://plugins.jenkins.io/code-coverage-api)

Implementations:
* [GitHub Checks Plugin](https://plugins.jenkins.io/github-checks)
```

> ## Consumers Guide

Reviewer: I think the guide belongs in its own file in the docs directory.

I would consider moving pipeline above this, as users will want this.

### A Simple Example

Imagine that you want to publish a check with some details of a Jenkins build. 
You need to first construct a `ChecksDetails` object:

```
ChecksDetails details = new ChecksDetailsBuilder()
 .withName("Jenkins CI")
 .withStatus(ChecksStatus.COMPLETED)
 .withConclusion(ChecksConclusion.SUCCESS)

>  .withDetailsURL("https://ci.jenkins.io")

Reviewer: probably should include in the example how to get the url i.e. from the display url api?

 .withCompletedAt(LocalDateTime.now(ZoneOffset.UTC))
 .build();
```

> Then you can create a publisher based on a Jenkins `run` to publish the check you just constructed:

Reviewer: ```suggestion
Then you can create a publisher based on a Jenkins `Run` to publish the check you just constructed:
```

```
ChecksPublisher publisher = ChecksPublisher.fromRun(run);
publisher.publish(details);
```

The publisher returned is based on the implementations you installed on your Jenkins instance.

### Checks Parameters

The checks are highly customized by consumers due to a number of optional parameters provided.
Consumers can set these parameters through the checks models:

> - `ChecksDetails`: the top-level model of a check, including all other models and some basic parameters like status and conclusion;

Reviewer: ```suggestion
- `ChecksDetails`: the top-level model of a check, including all other models and some basic parameters like status and conclusion
```

> - `ChecksOutput`: the output of a check, providing some displayed details like title, summary, URL, and code annotations;

Reviewer: ```suggestion
- `ChecksOutput`: the output of a check, providing some displayed details like title, summary, URL, and code annotations
```

- `ChecksAnnotation`: a code annotation of a check, providing a specific comment for one or multiple lines of code;

> - `ChecksImage`: an image of a check, providing an intuitive graph for the build like issues trend, coverage chart;

Reviewer: ```suggestion
- `ChecksImage`: an image of a check, providing an intuitive graph for the build like issues trend, coverage chart, must be a public URL that GitHub can fetch from
```

> - `ChecksAction`: an action of a check, providing further actions to be performed by users, like rerun a Jenkins build; 

Reviewer: ```suggestion
- `ChecksAction`: an action of a check, providing further actions to be performed by users, like rerun a Jenkins build
```

### Checks Publishers

The publishers are created through the static factory method (`fromRun` or `fromJob`) of `ChecksPublisherFactory`.
The factory will iterate all available implementations of the `ChecksPublisher` in order to find the suitable publisher for the Jenkins `Run` or `Job`.

## Embedded Features

### Build Status Check

> ![GitHub Status](docs/images/github-status.png)

Reviewer: should also include a screenshot showing the checks on ci.jenkins.io, probably near the top of the README

Author: I prefer to add that screenshot in the GitHub implementation, but I think I should add a section (and a link to the actions tab of this repo) after the known implementations to tell them where to take a look at the results.

Implementations:
* [GitHub Checks Plugin](https://plugins.jenkins.io/github-checks)

> This plugin, along with it's [GitHub implementation](https://plugins.jenkins.io/github-checks), has been installed on [ci.jenkins.io](https://ci.jenkins.io/Plugins) to help maintain over 1500 of Jenkins plugins. You can take a look at the [action](https://github.com/jenkinsci/checks-api-plugin/runs/1025532156) for this repository or other plugin repositories under [Jenkins organization](https://github.com/jenkinsci) for the results.

Reviewer: ```suggestion
This plugin, along with it's [GitHub implementation](https://plugins.jenkins.io/github-checks), has been installed on [ci.jenkins.io](https://ci.jenkins.io/Plugins) to help maintain over 1500 Jenkins plugins. You can take a look at the [action](https://github.com/jenkinsci/checks-api-plugin/runs/1025532156) for this repository or other plugin repositories under [Jenkins organization](https://github.com/jenkinsci) for the results.
```
